LibraBFT/Impl/Consensus/ConsensusTypes/Properties/QuorumCert.agda

```diff
@@ -0,0 +1,97 @@
+{- Byzantine Fault Tolerant Consensus Verification in Agda, version 0.9.
+
+   Copyright (c) 2021, Oracle and/or its affiliates.
+   Licensed under the Universal Permissive License v 1.0 as shown at https://opensource.oracle.com/licenses/upl
+-}
+ 
+open import LibraBFT.Base.Types
+open import LibraBFT.Base.KVMap as Map
+open import LibraBFT.Hash
+import      LibraBFT.Impl.Consensus.ConsensusTypes.Properties.VoteData as VoteDataProps
+import      LibraBFT.Impl.Consensus.ConsensusTypes.VoteData            as VoteData
+import      LibraBFT.Impl.Types.LedgerInfoWithSignatures               as LedgerInfoWithSignatures
+import      LibraBFT.Impl.Types.Properties.LedgerInfoWithSignatures    as LedgerInfoWithSignaturesProps
+open import LibraBFT.Impl.Consensus.ConsensusTypes.QuorumCert          as QuorumCert
+open import LibraBFT.ImplShared.Base.Types
+open import LibraBFT.ImplShared.Consensus.Types
+open import LibraBFT.ImplShared.Util.Util
+open import LibraBFT.Prelude
+open import Optics.All
+
+module LibraBFT.Impl.Consensus.ConsensusTypes.Properties.QuorumCert (self : QuorumCert) (vv : ValidatorVerifier) where
+  voteHash = hashVD (self ^∙ qcVoteData)
+
+  record rnd≡0Props : Set where
+    field
+        par≡cert : self ^∙ qcParentBlock    ≡ self ^∙ qcCertifiedBlock
+        cert≡li  : self ^∙ qcCertifiedBlock ≡ self ^∙ qcLedgerInfo ∙ liwsLedgerInfo ∙ liCommitInfo
+        noSigs   : Map.kvm-size (self ^∙ qcLedgerInfo ∙ liwsSignatures) ≡ 0
+  open rnd≡0Props
+
+  record rnd≢0Props : Set where
+    field
+      sigProp : LedgerInfoWithSignatures.verifySignatures (self ^∙ qcLedgerInfo) vv ≡ Right unit
+      vdProp  : VoteDataProps.Contract (self ^∙ qcVoteData)
+  open rnd≢0Props
+
+  rnd≡0 = self ^∙ qcCertifiedBlock ∙ biRound ≡ 0
+
+  record Contract : Set where
+    field
+      lihash≡ : self ^∙ qcSignedLedgerInfo ∙ liwsLedgerInfo ∙ liConsensusDataHash ≡ voteHash
+      rnd0    :   rnd≡0 → rnd≡0Props
+      ¬rnd0   : ¬ rnd≡0 → rnd≢0Props
```

Comment:
I think the contract would be more usefully stated if it contained a field with the decision for whether the round is 0, and a field whose type is dependent on that decision which evaluates to `rnd≡0Props` if it is `yes` and `rnd≢0Props` otherwise. That way, the user of this contract does not need to manually inspect the expression `self ^∙ qcCertifiedBlock ∙ biRound ≟ 0` -- just the field.

Reply:
Good idea, done, thanks.